modules/ginas/app/ix/ginas/utils/UniiLikeGenerator.java

```diff
@@ -0,0 +1,89 @@
+package ix.ginas.utils;
+
+import ix.ginas.controllers.v1.SubstanceFactory;
+import ix.ginas.models.v1.Substance;
+
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public abstract class UniiLikeGenerator extends RandomAlphaNumericIDGenerator<Substance> implements NamedIdGenerator<Substance,String>{
+
+    public static char[] alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ".toCharArray();
+    //GSRS-1658 can not contain only digit characters and "E" characters (`^[0-9E]*$`)
+    private static Pattern SCI_NOTATION_PATTERN = Pattern.compile("^[0-9E]+$");
+    //GSRS-1658 can not contain month abbreviations which cause problems in excel
+    private static Set<String> MONTHS = new HashSet<>(Arrays.asList("JAN", "FEB", "MAR", "APR", "MAY", "JUN", "JUL", "AUG", "SEP", "OCT", "NOV", "DEC"));
+    private static List<Character> CONFUSING_CHARS_LIST = Arrays.asList('0', 'O', '1', 'I');
+
+    private static char[] CONFUSING_CHARS_ARRAY;
+    static{
+        CONFUSING_CHARS_ARRAY = new char[CONFUSING_CHARS_LIST.size()];
```

Comment:
I have an aversion to static blocks. I may have rewritten this as a cachedSupplier, but it's a matter of preference.